### DISCUSSION TITLE
Group by doesn't show values for formula columns

### DISCUSSION THREAD

**Discussion by mpodshivalin** [Budibase/budibase#2271] [Help] [answered]
**Describe the bug**
Group by doesn't show values for formula columns, it shows an empty string

**To Reproduce**
1. Create a table, e.g. "Contracts"
2. Create Payment Date column (datetime), and a Price column (number)
3. Create an auto column Payment Month with a formula `{{#if Payment Date}}{{date Payment Date "MM.YYYY" }}{{/if}}` (shows only MM.YYYY part of payment)
4. Create a View, which calculates price by sum, and groups by Payment month

It will show group as empty:
![image](https://user-images.githubusercontent.com/16121630/127861219-86cdc5a8-696d-4608-a17d-fe854a62c4b2.png)

**Expected behavior**
It should show Payment month, e.g. 07.2021, 08.2021, etc.
That way it would be possible to make an aggregated statistics about monthly sales and show it in graps. Or am I doing something wrong?

**Desktop (please complete the following information):**
 - OS: Fedora 34
 - Browser Firefox 90
 - Latest Budibase (docker-compose version)

**Comment by mpodshivalin** [comment]
I've noticed that if you go to the row and explicitly save it, group by will have a value. So the problem may be narrowed down to the situation when the formula column was added after the initial row was created

**Comment by mike12345567** [comment]
hi @mpodshivalin - this is a feature that Budibase is not currently able to support. The implementation of formula fields is such that they are calculated when the row is retrieved, that way the formula value will always be up to date with the row when being fetched.

The view/calculate/group by functionality is carried out within the internal database, which does not know about the value of the formula field (as it shouldn't have been stored along with the row). You've actually found a slightly different bug, by saving the row again it seems the system has stored an actual value for the formula, but this will not stay up to date, if you change one of the values that the formula is based on (such as something it is related to) the formula value will not be updated.

I'll fix the issue of the formula values being stored and I'll make it so that formula fields cannot be picked as part of the view, as it isn't something that we can easily support.

**Comment by mpodshivalin** [comment]
Hmm so there can be no easy solution for making monthly sales reports (or making other aggregations)? 

For me, the biggest plus of Budibase is quickly making CRUD interfaces - something that consumes a lot of time (especially when requirements are frequently changing). 95% of these basic needs can be covered to Budibase functionality. However, what can I do if I want a custom report from the data I have in Budibase? E.g. I want to cover 100% of client's requirements, so I want to code remaining 5% of functionality, by making direct SQL queries or by other ways. The ultimate goal in my opinion is to be able to make and support the app in Budibase, not making a prototype in Budibase and then coding from scratch.
Basics are good and very well developed, but what can we do when a more advanced functionality is needed (it's a small portion of a big application, but can be a deal breaker)?

**Comment by mike12345567** [comment, marked as answer]
Hi @mpodshivalin this is something that is being looked at over the next while, theres a few things you can already do which can provide much deeper functionality and there is some work coming in the future as well.
1. Right now you can use helpers in our bindings, these are called Handlebars helpers, you can find some information about the functionality supported here: https://github.com/Budibase/handlebars-helpers
2. We are currently working on introducing JS everywhere, meaning that everywhere bindings are supported right now it will be possible to write a Javascript function to transform the data - making it more developer friendly.
3. If you are using an external SQL data source it is already possible to write custom SQL queries into this data, allowing you to produce very specific data formats through SQL. If you are using Postgres/MySQL you will get a very similar experience to what you have today with internal tables, but with the ability to write SQL queries as well.
4. In terms of filtering data, you have the data provider filter functionality, available for internal and external data sources - with the help of bindings you can retrieve and filter your data in pretty much any format.
5. We are also planning to introduce JS based data transformers, this will allow you to essentially write a JS function which the data will be put through on the way out of the API, transforming it in whatever way you desire. This is similar to what views provide today, but it won't be as performant (as it won't be done in database) but comes with the benefit of having access to all Budibase features.

I hope this helps explain some of the ways we're trying to make data filtering and transformation more customisable over the next while.

**Comment by mike12345567** [comment]
Converting this to discussion as a fix for original issue is being merged.

**Reply by mpodshivalin** [reply]
Thank you for such a detailed description!
Introducing JS everywhere seems like a good solution, so that non-standard things can be implemented without a need of rewrite.

> If you are using an external SQL data source it is already possible to write custom SQL queries into this data, allowing you to produce very specific data formats through SQL. If you are using Postgres/MySQL you will get a very similar experience to what you have today with internal tables, but with the ability to write SQL queries as well.

I initially thought about that. I've tried to use external data source (MySQL). But it seemed that it is not a similar experience to internal tables, 2 main points being:
1. I'm not able to create tables? I have to create them inside a database and then fetch it?
2. What about relationships? I rely on relationships being human-readable. If I specify a foreign key in MySQL, I still see IDs in tables.

Using external data source would be perfect, since it's a standard SQL database where I can make complex reports. But it seems to me that I can't create tables in Budibase and can't create relationships. Or am I missing something?

**Reply by mike12345567** [reply]
Hi @mpodshivalin on your questions around mySQL -
1. For now yes you need to create the tables externally, this is because table creation/management in SQL is quite complicated, creation differs between dialects of SQL and its hard for us to really fully understand your data requirements. We felt it was easier to define the tables with SQL as this gives the user complete control, something that we wouldn't be able to guarantee. We may visit this in the future and look at how Budibase could manage tables as well as data in SQL, but for now it is not on the roadmap.
2. As of the July release (last week or so) relationships are now manageable in MySQL and Postgres. In your data source management page, where you fetch your SQL tables you can specify the relationships between tables, telling Budibase about your foreign keys or junction tables (for many to many relationships). If this is done then you will still see the IDs, but it will also create a relationship column like you get with internal tables.
